tempodb/encoding/common/bloom.go

```diff
@@ -3,40 +3,53 @@ package common
 import (
 	"bytes"
 
-	"github.com/grafana/tempo/pkg/util"
 	"github.com/willf/bloom"
+
+	"github.com/grafana/tempo/pkg/util"
 )
 
-const shardNum = 10
+const legacyShardCount = 10
 
 type ShardedBloomFilter struct {
 	blooms []*bloom.BloomFilter
 }
 
-func NewWithEstimates(n uint, fp float64) *ShardedBloomFilter {
-	b := &ShardedBloomFilter{
-		blooms: make([]*bloom.BloomFilter, shardNum),
+// NewBloom creates a ShardedBloomFilter
+func NewBloom(fp float64, shardSize, estimatedObjects uint) *ShardedBloomFilter {
+	// estimate the number of shards needed. an approximate value is enough
+	var shardCount uint
+	var kPerBloom uint
+	for {
```

Comment:
https://github.com/bits-and-blooms/bloom/blob/master/bloom.go#L115

We should be able to calculate the shard size directly.  We don't need this loop.

Reply:
I am using exactly the same function `bloom.EstimateParameters`. Do you mean the following?
```
m, k := bloom.EstimateParameters(estimatedObjects, fp)

shardCount := m / (shardSize * 8)
```
I'm not sure if it works that way, does a smaller bloom with the same `k` have the same `fp`? I'll have to review that

Reply:
I believe it's possible:

https://hur.st/bloomfilter/

Reply:
It should err on the side of higher precision by rounding up the number of shards, i.e. `ceil(m / (shardSize * 8))`.  1/10th m with 1/10th objects is very close to the same FP, should be fine as long as shard count is rounded up.

Reply:
Yep, will round up 👍